src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -177,26 +183,25 @@ private bool CompleteCore()
                 _waiters = null;
             }
 
-            Task.Run(() =>
+            if (existingWaiters?.Count > 0)
             {
-                if (existingWaiters?.Count > 0)
+                // cancel waiters.
+                // NOTE: AsyncQueue has an invariant that 
+                //       the queue can either have waiters or items, not both
+                //       adding an item would "unwait" the waiters
+                //       the fact that we _had_ waiters at the time we completed the queue
+                //       guarantees that there is no items in the queue now or in the future, 
+                //       so it is safe to cancel waiters with no loss of diagnostics
+                Debug.Assert(this.Count == 0, "we should not be cancelling the waiters when we have items in the queue");
+                foreach (var tcs in existingWaiters)
                 {
-                    // cancel waiters.
-                    // NOTE: AsyncQueue has an invariant that 
-                    //       the queue can either have waiters or items, not both
-                    //       adding an item would "unwait" the waiters
-                    //       the fact that we _had_ waiters at the time we completed the queue
-                    //       guarantees that there is no items in the queue now or in the future, 
-                    //       so it is safe to cancel waiters with no loss of diagnostics
-                    Debug.Assert(this.Count == 0, "we should not be cancelling the waiters when we have items in the queue");
-                    foreach (var tcs in existingWaiters)
-                    {
-                        tcs.SetResult(default);
-                    }
+                    Debug.Assert(tcs.Task.CreationOptions.HasFlag(TaskCreationOptions.RunContinuationsAsynchronously));
+                    tcs.TrySetResult(default);
```

Comment:
Hmm, this maintains the interesting behavior of `AsyncQueue<T>.Dequeue`. Essentially there is no way to differentiate a completed `Task<T>` on whether it completed via an `Enqueue` operation or if it completed via `AsyncQueue<T>.Complete`. Not sure there is a good answer here though except for changing the fundamental contract of the type though :shrug:

Reply:
The return types are different. In this case, `default` is actually `default(Optional<T>)`, which has the `HasValue` set to false meaning the `Dequeue` did not complete.